### ISSUE TITLE
Delay the decision of which secret to use

### ISSUE THREAD

**Issue by ghost** [funcool/buddy-auth#73]
Hello Andrey, I ran into the use case where they JWS my API receives might be signed by different public keys each time and I have to read the token's header claim **kid** (or **jku**) to figure out against which key I need to verify the signature (the `:secret` param of the jws backend: https://github.com/funcool/buddy-auth/blob/master/src/buddy/auth/backends/token.clj#L36). I was wondering if it's a good idea to extend the jws backend implkeyementation to receive a function as the `:secret` in which case the JWS header is passed to it and returns the appropriate secret.

A real example from my project:
When my web service starts I query my authentication service to receive all the public keys available to verify JWS' signatures:

```clj
(defstate keys
  "Holds the public keys used for verifying JWS' signatures"
  :start (...))
```

With the current implementation, when defining the middleware of my app the only choice I have is to pick one of the keys defined above:

```clj
defn base-middleware
  "Application base middlewares."
  [handler]
  (let [secret (first keys)]
    (-> handler
        ...
        (wrap-authentication (app.auth/jws-backend secret))
        ...
        )))
```

But what I would like is to be able to choose a key depending on the claims of the JWS itself, for example:

```clj
defn base-middleware
  "Application base middlewares."
  [handler]
  (let [secret (fn [header]
                      (get-key-by-id keys (:kid header)))]
    (-> handler
        ...
        (wrap-authentication (app.auth/jws-backend secret))
        ...
        )))
```

What do you think?

**Comment by snorremd** [comment]
I second @anler's suggestion. It would be really useful to be able to select the secret key validation time based on the incoming JWT header (e.g. key id). I think this is a fairly common use case. My use case specifically has to do with openid connect style authorization flows, where the identity server exposes a list of one or more JWKs (json web keys) and I need to select the key to use for signature validation based on the key id supplied in the JWT header.

Edit: Supporting a secret-function would of course not have to remove the possibility of supplying a key when creating the backend.

**Comment by niwinz** [comment]
It is already implemented, just bump the buddy sign to 3.0.0 that uses the new `IKeyProvider` protocol: https://github.com/funcool/buddy-sign/blob/master/src/buddy/sign/util.clj#L30

By default right now if you pass a function as the secret, this function will be executed with a header as argument and you should return the correct key.
